src/plugins/discover/public/application/main/hooks/use_text_based_query_language.ts

```diff
@@ -78,30 +78,44 @@ export function useTextBasedQueryLanguage({
             prev.current = { columns: firstRowColumns, query };
             nextColumns = firstRowColumns;
           }
+
           if (firstRowColumns && initialFetch) {
             prev.current = { columns: firstRowColumns, query };
           }
         }
         const indexPatternFromQuery = getIndexPatternFromSQLQuery(query.sql);
-        const dataViewObj = dataViewList.find(({ title }) => title === indexPatternFromQuery);
+        let dataViewObj = dataViewList.find(({ title }) => title === indexPatternFromQuery);
+
+        // no dataview found but the index pattern is valid
+        // create an adhoc instance instead
+        if (!dataViewObj) {
+          dataViewObj = await dataViews.create({
+            title: indexPatternFromQuery,
+          });
 
-        if (dataViewObj) {
-          // don't set the columns on initial fetch, to prevent overwriting existing state
-          const addColumnsToState = Boolean(
-            nextColumns.length && (!initialFetch || !stateColumns?.length)
-          );
-          // no need to reset index to state if it hasn't changed
-          const addDataViewToState = Boolean(dataViewObj.id !== index);
-          if (!addColumnsToState && !addDataViewToState) {
-            return;
+          if (dataViewObj.fields.getByName('@timestamp')?.type === 'date') {
+            dataViewObj.timeFieldName = '@timestamp';
+          } else if (dataViewObj.fields.getByType('date')) {
```

Comment:
This is a great change ... data views are becoming an internal implementation detail, users can focus on what matters: finding the data of their desire
While testing I encountered, that switching to a data view without a timefield doesn't work. With the following change it should:
```suggestion
          } else if (dataViewObj.fields.getByType('date').length) {
```

Reply:
Great find Matthias. Fixed for Lens and Discover https://github.com/elastic/kibana/pull/147423/commits/f88b142da3b5b46b273f33bf31ef66bb7ad464a0